### PR TITLE
Use common module for subroutine and entry function 

### LINKED ISSUE #2380: Original variables do not get updated when updating variable inside entry function
Ahh, this happens because we do not update the parent variables on calling entry functions.

```fortran
      SUBROUTINE dinvr(x)
      IMPLICIT NONE
      DOUBLE PRECISION x
      DOUBLE PRECISION big,small,zx,zy,zz, zsmall, zbig
      LOGICAL qcond
      INTRINSIC abs,max,min
      LOGICAL qxmon

      qxmon(zx,zy,zz) = zx .LE. zy .AND. zy .LE. zz

      qcond = .NOT. qxmon(small,x,big)

      print *, "small = ", small
      print *, "x = ", x
      print *, "big = ", big

      IF (qcond) error stop
      print *, 'qcond = ', qcond

      entry distinv(zsmall, zbig)

      small = 0.0d0
      big = 1.0d0

      return


      end subroutine

      program main
      implicit none
      double precision :: x
      x = 0.5d0

      call distinv(0.0d0, 1.0d0)

      call dinvr(x)
      end program
```

> GFortran

```console
% gfortran a.f90 && ./a.out
 small =    0.0000000000000000     
 x =   0.50000000000000000     
 big =    1.0000000000000000     
 qcond =  F
```

> LFortran 

```console
% lfortran a.f90 --no-warnings
small =  1.43349365507895023e-297
x =  5.00000000000000000e-01
big =  4.94065645841246544e-324
ERROR STOP
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #2380

### CODE REVIEW

**Comment by Pranavchiku** [comment]
This https://github.com/lfortran/lfortran/issues/2380#issue-1884358720 does not work yet.

File: tests/reference/asr-entry_02-b8ee2e2.stdout [hunk 1]
```diff
@@ -191,14 +191,26 @@
                     .false.
                     ()
                 ),
+            x_saved_variables:
```
Comment (Pranavchiku) [inline tests/reference/asr-entry_02-b8ee2e2.stdout:194]: I reviewed this updated references, it looks good to me.

**Comment by Pranavchiku** [comment]
On checking scipy/special/cdflib/* manually, I got
```console
$ lfortran -c /Users/pranavchiku/repos/scipy/scipy/special/cdflib/dinvr.f -o /Users/pranavchiku/repos/scipy/scipy/special/cdflib/dinvr.o --fixed-form --implicit-typing --implicit-interface --generate-object-code --use-loop-variable-after-loop --all-mangling --mangle-underscore --bindc-mangling
semantic error: Assign target needs to be a variable.
   --> /Users/pranavchiku/repos/scipy/scipy/special/cdflib/dinvr.f:100:7
    |
100 |       ASSIGN 10 TO i99999
    |       ^^^^^^^^^^^^^^^^^^^ 


Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

**Comment by Pranavchiku** [comment]
Yes, this happened because we had a `SAVE` declared. This resulted in declaring of every variable inside the module and it was an external symbol, I got it fixed. Now, I am facing another one, I will resove this and push the changes.

**Comment by Pranavchiku** [comment]
Minimal code for SciPy failure

```fortran
subroutine id_frand(n,r)
implicit none
integer n
real*8 r(n)
save
entry id_frandi()
return
end
```

This happens because `symbol_duplicator.duplicat_symbol)` is not able to duplicate type of array dimension variable correctly.

**Comment by Pranavchiku** [comment]
I have a fix for this ready but, there are few other scipy failures. I am trying to fix those, looks like this PR affects almost every case of SciPy, I will need some more time to get this PR ready to review.

**Comment by certik** [comment]
Here is a good test case that works:
```fortran
SUBROUTINE dinvr(x)
IMPLICIT NONE
DOUBLE PRECISION x
DOUBLE PRECISION big,small,zx,zy,zz, zsmall, zbig
LOGICAL qcond
INTRINSIC abs,max,min
LOGICAL qxmon
SAVE

qxmon(zx,zy,zz) = zx .LE. zy .AND. zy .LE. zz

qcond = .NOT. qxmon(small,x,big)

print *, "small = ", small
print *, "x = ", x
print *, "big = ", big

print *, 'qcond = ', qcond
IF (qcond) error stop

return

entry distinv(zsmall, zbig)

small = zsmall
big = zbig

end subroutine

program entry_10
implicit none
double precision :: x
x = 0.5d0

call distinv(0.0d0, 1.0d0)

call dinvr(x)
end program
```

**Comment by Pranavchiku** [comment]
Closing with reference to #2676, #2680, #2564.

**Pull request closed without merging**